### ISSUE TITLE
nan loss late in training

### ISSUE THREAD

**Issue by wsnoble** [Noble-Lab/casanovo#412]
## Describe the Issue
I tried to train a Casanovo model, but the losses turn into NaNs after ~450k batches.

## Steps To Reproduce
Here is the command line:

```
casanovo train /net/noble/vol1/home/gstrau2/data/downsample_splits/multienzymeds.train.mgf --validation_peak_path /net/noble/vol1/home/gstrau2/data/downsample_splits/multienzymeds.val.mgf --config casanovo.yaml --output_dir downsample
```

Config file is [here](https://github.com/user-attachments/files/18030014/casanovo.yaml.txt).

## Expected Behavior
I don't expect to see NaN's!  

## Terminal Output (If Applicable)
Provide any applicable console output in between the tick marks below.

Here is what I see in the log file 

```
2024-12-02 18:22:21,587 INFO [casanovo/MainProcess] model._log_history : 50000	nan	1.068294
2024-12-02 22:23:49,765 INFO [casanovo/MainProcess] model._log_history : 98010	0.816915	nan
2024-12-02 22:37:34,798 INFO [casanovo/MainProcess] model._log_history : 100000	nan	0.848568
2024-12-03 02:34:03,041 INFO [casanovo/MainProcess] model._log_history : 147015	0.656524	nan
2024-12-03 02:52:47,707 INFO [casanovo/MainProcess] model._log_history : 150000	nan	0.684293
2024-12-03 06:43:52,442 INFO [casanovo/MainProcess] model._log_history : 196020	0.579432	nan
2024-12-03 07:07:33,894 INFO [casanovo/MainProcess] model._log_history : 200000	nan	0.622089
2024-12-03 10:53:37,664 INFO [casanovo/MainProcess] model._log_history : 245025	0.536403	nan
2024-12-03 11:22:33,267 INFO [casanovo/MainProcess] model._log_history : 250000	nan	0.577466
2024-12-03 15:03:55,538 INFO [casanovo/MainProcess] model._log_history : 294030	0.502367	nan
2024-12-03 15:37:41,791 INFO [casanovo/MainProcess] model._log_history : 300000	nan	0.536349
2024-12-03 19:14:21,014 INFO [casanovo/MainProcess] model._log_history : 343035	0.472080	nan
2024-12-03 19:53:04,503 INFO [casanovo/MainProcess] model._log_history : 350000	nan	0.505742
2024-12-03 23:24:02,820 INFO [casanovo/MainProcess] model._log_history : 392040	0.443729	nan
2024-12-04 00:07:41,937 INFO [casanovo/MainProcess] model._log_history : 400000	nan	0.475907
2024-12-04 03:33:43,388 INFO [casanovo/MainProcess] model._log_history : 441045	0.416855	nan
2024-12-04 04:22:18,173 INFO [casanovo/MainProcess] model._log_history : 450000	nan	0.448906
2024-12-04 07:40:37,601 INFO [casanovo/MainProcess] model._log_history : 490050	nan	nan
2024-12-04 08:33:17,494 INFO [casanovo/MainProcess] model._log_history : 500000	nan	nan
2024-12-04 11:45:31,374 INFO [casanovo/MainProcess] model._log_history : 539055	nan	nan
2024-12-04 12:42:56,594 INFO [casanovo/MainProcess] model._log_history : 550000	nan	nan
2024-12-04 15:50:08,945 INFO [casanovo/MainProcess] model._log_history : 588060	nan	nan
2024-12-04 16:52:47,124 INFO [casanovo/MainProcess] model._log_history : 600000	nan	nan
2024-12-04 19:55:26,912 INFO [casanovo/MainProcess] model._log_history : 637065	nan	nan
2024-12-04 21:02:46,496 INFO [casanovo/MainProcess] model._log_history : 650000	nan	nan
2024-12-05 00:00:18,987 INFO [casanovo/MainProcess] model._log_history : 686070	nan	nan
2024-12-05 01:12:35,198 INFO [casanovo/MainProcess] model._log_history : 700000	nan	nan
2024-12-05 04:05:47,228 INFO [casanovo/MainProcess] model._log_history : 735075	nan	nan
2024-12-05 05:22:50,959 INFO [casanovo/MainProcess] model._log_history : 750000	nan	nan
2024-12-05 08:10:46,844 INFO [casanovo/MainProcess] model._log_history : 784080	nan	nan
```

## Environment:
- OS: Ubuntu
- Casanovo Version: 4.2.2.dev107+g17bb3f2
- Hardware Used (CPU or GPU, if GPU also GPU model and CUDA version): n025

[casanovo_20241202135518.log.txt](https://github.com/user-attachments/files/18030028/casanovo_20241202135518.log.txt)

**Comment by wsnoble** [comment]
OK, I've got a dataset that yields this nan problem immediately, rather than a day later.  The command line is basically the same as above, and the yaml file is attached.  The log file gives nan for both training and test loss immediately.

It seems like, at the very least, we should detect a nan loss and terminate when it happens.

`casanovo train ../2024-10-11perturb/perturbed.train.mgf --validation_peak_path ../2024-10-11perturb/perturbed.val.mgf --config casanovo.yaml --output_dir perturbed-2`

'''
2024-12-12 11:19:57,727 INFO [casanovo/MainProcess] model._log_history : Step	Train loss	Valid loss	
2024-12-12 11:19:57,727 INFO [casanovo/MainProcess] model._log_history : 49005	nan	nan
2024-12-12 11:23:50,999 INFO [casanovo/MainProcess] model._log_history : 50000	nan	nan
2024-12-12 13:28:11,614 INFO [casanovo/MainProcess] model._log_history : 98010	nan	nan
2024-12-12 13:34:36,535 INFO [casanovo/MainProcess] model._log_history : 100000	nan	nan
```
[casanovo_20241212090720.log.txt](https://github.com/user-attachments/files/18117789/casanovo_20241212090720.log.txt)

**Comment by wsnoble** [comment]
This seems to depend on the Cuda version.  The following upgrade fixed the problem:

```
Before:
| NVIDIA-SMI 560.28.03              Driver Version: 560.28.03      CUDA Version: 12.6     |

After:
| NVIDIA-SMI 560.35.05              Driver Version: 560.35.05      CUDA Version: 12.6     |
```

**Comment by wsnoble** [comment]
For people who run into this bug, I think it would be good if Casanovo would either (1) stop when nan is reached, or (2) skip updating the weights and try computing a new set of gradients with a lower learning rate.  The current behavior is to happily continue generating nan losses forever, which is not super useful.
